pkg/audit/filter.go

```diff
@@ -69,3 +69,18 @@ func (aw *wrapWriter) Hijack() (net.Conn, *bufio.ReadWriter, error) {
 	}
 	return nil, nil, errors.New("the ResponseWriter doesn't support the Hijacker interface")
 }
+
+func (aw *wrapWriter) CloseNotify() <-chan bool {
+	if cn, ok := aw.ResponseWriter.(http.CloseNotifier); ok {
+		return cn.CloseNotify()
+	}
+	panic("the ResponseWriter doesn't support the CloseNotifier interface")
```

Comment:
I wonder if it's a common paradigm  - to panic on api filter?

Reply:
I found panic, convert directly and have two different writer, k8s use the support two writer, but if we use that way, when the not implement flush interface responseWriter try to call the flush function will still block, and no error, so I prefer the panic.   
https://github.com/mholt/caddy/blob/master/caddyhttp/httpserver/responsewriterwrapper.go
https://github.com/go-macaron/macaron/blob/master/response_writer.go
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/endpoints/filters/audit.go

Reply:
@alena1108 As we found a reference in `https://github.com/mholt/caddy/blob/master/caddyhttp/httpserver/responsewriterwrapper.go#L55`, we would prefer panic if type assertion fails.
We also found what Kubernetes does in `https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/endpoints/filters/audit.go`. 
Its audit log filter will generate a `ResponseWriter` implements the interfaces `http.Flusher`, `http.Hijacker` and `http.CloseNotifier` if the original `ResponseWriter` also does. 
But It also doesn't do anything if the original `ResponseWriter` doesn't implement them. So it won't show logs/errors when it gets into the same situation. So we want to introduce panic to shows us what goes wrong.
@aiwantaozi Please correct me if I misunderstand anything.

Reply:
Does this cause us to log a panic stacktrace in our logs?

Reply:
@aiwantaozi @orangedeng logging the message in non-disruptive manner + returning from the message should be sufficient enough. Panic is way too big

Reply:
@cjellick will not cause panic stacktrace every time someone deletes a namespace from kubectl. It will convert fail when upstream wrapper responseWriter but not support CloseNotifier and Flusher, when the client(kubectl) tries to use these function then get blocked.

Reply:
@alena1108 update to use log the error